### ISSUE TITLE
AnalysisContext.RegisterCodeBlockStartAction regressed: called twice in the middle of method

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#2894]
The `RegisterCodeBlockStartAction` method used to be called at the start of each method, and is documented to do so. But after Dev14 RC, Roslyn regressed this and now the method is called twice in a single method, both *after* the method has begun being processed.

This caused a regression to be picked up by one of the unit tests for our own analyzer.

Repro steps:

1. Clone the [Repro project](https://github.com/AArnott/roslynbug2894.git)
1. Open the solution and run tests to verify that they all work. This is against Roslyn 1.0.0-rc2
1. Now update packages to the latest Roslyn nuget packages (or switch to the first commit in the repo, if you can restore those packages).
1. Re-run tests. You'll see that `InvokeVsSolutionAfterVerifyOnUIThread` fails.

**Comment by AArnott** [comment]
@mavasani volunteered to take this one.

**Comment by mavasani** [comment]
/cc @JohnHamby @srivatsn @ManishJayaswal @tmeschter 

I investigated the repro project and the issue seems that the analyzer that regressed (https://github.com/AArnott/roslynbug2894/blob/master/Microsoft.VisualStudio.Threading.Analyzers/VsServiceUsageAnalyzer.cs) is not correctly implemented, but worked up until rc2 because the analyzer driver wasn't fully utilizing parallelism in analyzer execution across symbols. Let me expand on this below.

The analyzer under question here has the below implementation of `Initialize` method.
```c#
        public override void Initialize(AnalysisContext context)
        {
            context.RegisterCodeBlockStartAction<SyntaxKind>(_ =>
                this.methodDeclarationNodes = this.methodDeclarationNodes.Clear()
            );
            context.RegisterSyntaxNodeAction(this.AnalyzeInvocation, SyntaxKind.InvocationExpression);
            context.RegisterSyntaxNodeAction(this.AnalyzeMemberAccess, SyntaxKind.SimpleMemberAccessExpression);
            context.RegisterSyntaxNodeAction(this.AnalyzeCast, SyntaxKind.CastExpression);
            context.RegisterSyntaxNodeAction(this.AnalyzeAs, SyntaxKind.AsExpression);
        }
```

Note that there is an instance field `methodDeclarationNodes` that is being cleared on the `RegisterCodeBlockStartAction` callback, and this field is used by callbacks on syntax node actions that are registered on same analysis context to manipulate the values on this dictionary, as well as report diagnostics based on its contents. This is an incorrect implementation.
This implementations means that if the analyzer's actions were invoked in the following order, your state will get corrupt at step 3:

1. `RegisterCodeBlockStartAction` on one of the declared method symbol => clear global state
2. Syntax node actions for first few statements of that body => manipulate global state
3. `RegisterCodeBlockStartAction` on another declared method => clear all the global state again
4. Alter between both these methods for executing syntax node actions on remaining nodes within the respective method bodies.

The global state in the instance field `methodDeclarationNodes` is corrupt after step 3. and the analyzer won't function as expected from thereon.

Another issue with the analyzer is that you are manipulating state as well as reporting diagnostics based on this collected state in the syntax node actions itself. This is neither functionally correct nor  a performant approach for writing such an analyzer. The analyzer driver is free to invoke syntax node actions for nodes within a method body in any order (it doesn't do so currently, but you should not rely on this implementation detail).
The correct approach would be to collect all state about syntax nodes within the body in your syntax node action callbacks and then report diagnostics based on the collected state in the code block end action, which is guaranteed to be invoked at the end of the code block.

I would say the correct implementation for this analyzer would be on the following lines, Sri/John/Tom can chime in as well:

```c#
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class VsServiceUsageAnalyzer : DiagnosticAnalyzer
    {
        ...

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterCodeBlockStartAction<SyntaxKind>(codeBlockContext =>
            {
                var codeBlockAnalyzer = new MyCodeBlockAnalyzer();

                // Manipulate state in these callbacks
                codeBlockContext.RegisterSyntaxNodeAction(codeBlockAnalyzer.AnalyzeInvocation, SyntaxKind.InvocationExpression);
                codeBlockContext.RegisterSyntaxNodeAction(codeBlockAnalyzer.AnalyzeMemberAccess, SyntaxKind.SimpleMemberAccessExpression);
                codeBlockContext.RegisterSyntaxNodeAction(codeBlockAnalyzer.AnalyzeCast, SyntaxKind.CastExpression);
                codeBlockContext.RegisterSyntaxNodeAction(codeBlockAnalyzer.AnalyzeAs, SyntaxKind.AsExpression);

                // Report diagnostics in this callback.
                codeBlockContext.RegisterCodeBlockEndAction(codeBlockAnalyzer.CodeBlockEndToReportDiagnostics);
            });            
        }

        private class MyCodeBlockAnalyzer
        {
            private ImmutableDictionary<SyntaxNode, ThreadingContext> methodDeclarationNodes = ImmutableDictionary<SyntaxNode, ThreadingContext>.Empty;
            public void AnalyzeInvocation(SyntaxNodeAnalysisContext context) { }
            public void AnalyzeMemberAccess(SyntaxNodeAnalysisContext context) { }
            public void AnalyzeCast(SyntaxNodeAnalysisContext context) { }
            public void AnalyzeAs(SyntaxNodeAnalysisContext context) { }
            public void CodeBlockEndToReportDiagnostics(CodeBlockAnalysisContext context) { }
        }
        ...
    }
```


Prior to rc2, our analyzer driver processed symbols for analyzer execution sequentially, i.e. it would go an execute all the symbol/syntax/code block actions for one symbol before moving to the next symbol. This was the biggest bottleneck for very slow build times with analyzers. Post rc2, we made our symbol processing + analyzer execution parallel (though the individual callbacks into the analyzer are guarded by a lock on a unique gate object created per-analyzer instance, so your analyzer callbacks need not be thread-safe).

I am resolving this bug as by design, please ping me, @JohnHamby or @tmeschter if you have further questions on this, we would be glad to help.

**Comment by AArnott** [comment]
Thanks, @mavasani both for your analysis and proposed fixes. I wondered what the thread safety obligations were for an analyzer. Another approach Roslyn might have taken was to instantiate the analyzer multiple times for each parallel track so that no instance had to be thread-safe itself. Did you consider that?

Also: when debugging the problem myself I noticed that 1 or 2 statements were processed *before* the  `RegisterCodeBlockStartAction` callback was invoked even the first time. How is that correct? Regardless of parallelism, shouldn't a method start callback have been invoked before analyzing any of its contents?

**Comment by tmeschter** [comment]
@AArnott This isn't a thread-safety issue per se, as we guarantee we don't call multiple actions on a given analyzer concurrently. Rather, your analyzer actions are not being called in the order that makes intuitive sense.

The ordering rules can be summarized as follows (@JohnHamby, please correct me if I'm wrong; I lifted these from an email you wrote up a while back):

1. For a given compilation, compilation start actions run first, and compilation end actions run last.
2. For a given code block, code block start actions run before syntax node actions, and syntax node actions run before code block end actions.
3. Otherwise, there are no ordering guarantees.

As for the issue where statements were being processed before the code block start action, my guess is that since the syntax actions were not registered within the code block start action we treated them as independent, but @mavasani or @JohnHamby would need to confirm that. @mavasani's suggested implementation correctly associates the actions and should work as expected.

**Comment by AArnott** [comment]
> my guess is that since the syntax actions were not registered within the code block start action we treated them as independent

If I register syntax actions in the code block start action, do I need to unregister them to avoid a memory leak or anything? How do I need to be aware of to avoid leaks in my analyzer?

**Comment by tmeschter** [comment]
No, you don't need to unregister them.

To avoid memory leaks, try to avoid saving any state in your analyzer (that is, don't store information in fields in the analyzer class). In particular, do not store Roslyn types in your analyzer (e.g. `Compilation`, `ISymbol`, etc.).

Also note that you cannot depend on all of your actions being called. For example, even though we call a code block start action, we do not guarantee that we call the corresponding code block end action. This is common in IDE scenarios, where we cancel analysis part of the way through due to user edits, and then restart. The implication is that there is no mechanism for deterministically releasing resources.

**Comment by srivatsn** [comment]
[This document](https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Analyzer%20Actions%20Semantics.md) has some more information about the semantics of analyzer actions.

**Comment by mavasani** [comment]
@AArnott @tmeschter Andrew has raised couple of distinct issues here, and I feel we should given more detailed documentation and/or samples around these. I also think Andrew's last comment is something that our defined semantics guarantee, but the implementation doesn't and that is a bug (I'll open a separate issue). All the comments made by Tom are perfectly right.

1. State management: How do I define/manipulate state in my stateful analyzers? What should be the lifetime/cleanup strategy for resources allocated during my analysis? Does the read/write to my resources need to be thread-safe?
2. Ordering guarantees on analyzer actions: The document Sri pointed to has a detailed explanation about this, but probably needs to be surfaced in a more visible way through doc comments and/or sample code snippets.

I'll open a separate GitHub tracking issue for adding more documentation/samples around these. The answer to them for your reference:

1. State management:
    1. Lifetime of analyzer: State stored on fields of analyzer can live very long, as analyzer instances can live for whole lifetime of analyzer host (i.e. in IDE for entire VS session for VSIX based analyzers). Only state that should be stored as fields:

        1. Immutable data that is computed in Initialize and re-used across all actions, is independent of compilation/symbols/syntax etc. and is purely some analyzer specific data.

        2. Mutable data that is compilation/symbol dependent but expensive to re-compute for every compilation that an analyzer might execute on (say for typing scenarios in IDE). You should be very careful in how this is done, as you can end up leaking compilation/symbols. Conditional weak tables are probably needed for this to be done correctly. This should be avoided as much as possible.
    
    2. State corresponding to compilation/code blocks etc: For stateful analysis, you must define a start action corresponding to the logical entity where you need stateful analysis. There are couple of recommended ways to then define/manipulate state:

        1. Locals defined in the start action: These are available to all the nested registered actions for the scope of analyzer execution on the corresponding logical entity.

        2. Define a new private nested type in analyzer that defines the nested actions, and the state can then be stored as fields on this type. Instantiate this type for every start action callback.

       Guideline here is to initialize state in the start action (or constructor of nested type for approach 2), manipulate it in the nested non-end actions, and finally use the accumulated state in the nested end action to report diagnostics.

2. Ordering guarantees: Theoretically, nesting is the best way to define your order-dependency on actions. You are guaranteed to have start actions run before any nested action, so this should be the most reliable implementation. However, as the semantics document suggests, we want to guarantee ordering for actions that logically belong inside an executable code block to run after start action, regardless of the point where syntax node actions were registered. So, the scenario you mentioned should work as per your expectation, and it is a bug that our analyzer driver doesn't guarantee this ordering for code block start and syntax node actions execution. I will file a separate issue, but this shouldn't block your analyzer.

**Comment by AArnott** [comment]
Thanks. I applied the suggested change to my analyzer and it passes now. :)

**Comment by mavasani** [comment]
Great!

**Comment by amcasey** [comment]
Is any verification required @Srivatsn?

**Comment by mavasani** [comment]
I think this one was by design.
